### DISCUSSION TITLE
Cannot start rustdesk as service (not ServiceProfiles Folder broken, always failed to create IPC service)

### DISCUSSION THREAD

**Discussion by lugt** [rustdesk/rustdesk#6640] [General]
### Bug Description

When I install rustdesk as service, the mouse will blink , process exit and start, show the wait ring and hide, then show and hide again.
I have to uninstall the software .(remote side)
uninstalling and reinstalling does not restore the situation.

the client program when started still attempts to connect to the pipe to contact the Service, which is not removed somehow during uninstallation process.  

PS. only occur when tried to install as a Service, and somehow that installation failed.

### How to Reproduce

Install as Service.

### Expected Behavior

Capable of proper uninstallation at least, does everything to restore the environment after installation.
Capable of installation as a Service.

### Operating system(s) on local side and remote side

both Windows 10 version 22H2, 19045.3570 -> Windows 10

### RustDesk Version(s) on local side and remote side

1.2.3 -> 1.2.3

### Screenshots

Closed in a split of a second

### Additional Context

```
[2023-12-08 11:27:42.718369 +08:00] INFO [src\core_main.rs:162] main start args:[]
[2023-12-08 11:27:42.838742 +08:00] DEBUG [libs\hbb_common\src\config.rs:475] Configuration path: C:\Users\lugt\AppData\Roaming\RustDesk\config\RustDesk_local.toml
[2023-12-08 11:27:42.856373 +08:00] DEBUG [libs\hbb_common\src\config.rs:475] Configuration path: C:\Users\lugt\AppData\Roaming\RustDesk\config\RustDesk.toml
[2023-12-08 11:27:42.866656 +08:00] DEBUG [libs\hbb_common\src\config.rs:475] Configuration path: C:\Users\lugt\AppData\Roaming\RustDesk\config\RustDesk_default.toml
[2023-12-08 11:27:42.870761 +08:00] DEBUG [libs\hbb_common\src\config.rs:475] Configuration path: C:\Users\lugt\AppData\Roaming\RustDesk\config\RustDesk2.toml
[2023-12-08 11:27:43.725921 +08:00] INFO [src\server.rs:409] server not started (will try to start): deadline has elapsed
[2023-12-08 11:27:43.727231 +08:00] ERROR [src\ipc.rs:293] Failed to start ipc server at path \\.\pipe\RustDesk\query: access denied. (os error 5)
[2023-12-08 11:27:43.727570 +08:00] ERROR [src\server.rs:374] Failed to start ipc: access denied. (os error 5)

```

**Comment by rustdesk** [comment]
It seems your system has special security settings. Can not create IPC on your system.

Answer from ChatGPT.

# Failed to create IPC on Windows, why?

> There are several reasons why you might be failing to create an IPC on Windows:
> 
> **1. IPC Service is not running:**
> 
> - The IPC service is responsible for enabling IPC communication in Windows. If it's not running, you won't be able to create an IPC.
> 
> **Solution:**
> 
> 1. Open the "Services" window.
> 2. Locate "Remote Procedure Call (RPC)".
> 3. Right-click on "Remote Procedure Call (RPC)" and select "Start".
> 
> **2. Insufficient IPC Permissions:**
> 
> - Creating an IPC requires specific permissions. If you don't have the necessary access, you won't be able to proceed.
> 
> **Solution:**
> 
> 1. Open the "Local Security Policy" window.
> 2. Expand "Local Policies"\"Security Options".
> 3. Locate "Allow local users to remotely connect using NTLM authentication (RPC)".
> 4. Ensure this policy is set to "Enabled".
> 
> **3. Lack of System Resources:**
> 
> - Creating an IPC consumes resources like memory and disk space. If your system is lacking, it might not be able to handle the request.
> 
> **Solution:**
> 
> 1. Open the "Task Manager".
> 2. Switch to the "Performance" tab.
> 3. Check the "Memory" and "Disk" usage.
> 
> **4. Third-party IPC Software Issues:**
> 
> - If you're using a third-party IPC software, there might be internal issues causing the problem. Contact the software developer for assistance.
> 
> **5. General System Issues:**
> 
> - Sometimes, temporary problems or glitches can hinder IPC creation.
> 
> **Solutions:**
> 
> - Restart your computer.
> - Create a new user account and try again.
> - Utilize Windows' built-in IPC tools.
> 
> **Additional Tips:**
> 
> - If you're still facing issues after trying the above solutions, consider consulting Windows technical support.
> - Double-check the specific error message you're receiving, as it might provide further clues about the cause of the problem.
> - If you're working with a specific software or program requiring IPC, consult its documentation or support channels for specific troubleshooting steps.
> 
> Remember, troubleshooting IPC issues can involve multiple factors. By systematically checking for the most common causes and applying the appropriate solutions, you should be able to resolve the problem and successfully create an IPC on your Windows system.

**Comment by deep-soft** [comment]
I think I have the same problem on Windows version above 21H2.
- The last nightly version works **OK** on Win 21H2, but **NOT** on Win10 22H2 or Win11.

- The last version which work **OK** on all Windows versions, mentioned above (Win10 21H2, Win10 22H2 and Win11), is the one compiled from commit 3f3bca7.
https://github.com/rustdesk/rustdesk/pull/6460

Message from log:
ERROR [src\server\connection.rs:1504] ipc to connection manager exit: expected
DEBUG [src\server\connection.rs:1507] SendError(CmErr("expected")), librustdesk::server::connection:src\server\connection.rs:1507:25

**Reply by 21pages** [reply]
Could you provide more logs of the abnormal part ?

**Reply by deep-soft** [reply]
sorry, was at work, i'll try to extract those logs.

**Reply by deep-soft** [reply]
Unfortunately I didn't keep the logs.
Tried to recreate same conditions but didn't find:
`ERROR [src\server\connection.rs:1504] ipc to connection manager exit: expected`

These are the logs, connection to Windows 11 Pro (with last Windows updates):
```---#--- RustDesk_r2023-12-14_23-45-45.log
[2023-12-14 23:45:16.196982 +02:00] INFO [src\core_main.rs:162] main start args:[]
[2023-12-14 23:45:16.198405 +02:00] INFO [src\server.rs:476] server not started (will try to start): The system cannot find the file specified. (os error 2)
[2023-12-14 23:45:16.200374 +02:00] INFO [src\ipc.rs:278] Started ipc server at path: \\.\pipe\RustDesk\query
[2023-12-14 23:45:16.205827 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk2.toml
[2023-12-14 23:45:16.207278 +02:00] INFO [src\lan.rs:30] lan discovery listener started
[2023-12-14 23:45:16.314370 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk_local.toml
[2023-12-14 23:45:16.318152 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk.toml
[2023-12-14 23:45:16.323449 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk_default.toml
[2023-12-14 23:45:18.242620 +02:00] INFO [libs\scrap\src\common\hwcodec.rs:382] Check hwcodec config, exit with: exit code: 0
[2023-12-14 23:45:19.085930 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk_default.toml
[2023-12-14 23:45:20.777950 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk_default.toml
[2023-12-14 23:45:42.105320 +02:00] INFO [src\platform\windows.rs:2176] Installing service...
[2023-12-14 23:45:42.250849 +02:00] INFO [src\rendezvous_mediator.rs:521] Direct server listening on: Some(Ok([::]:39878))
[2023-12-14 23:45:43.660377 +02:00] INFO [src\ipc.rs:344] Receive close message
[2023-12-14 23:45:43.665962 +02:00] DEBUG [src\platform\windows.rs:2262] Spawn new window

---#--- RustDesk_r2023-12-14_23-46-24.log
[2023-12-14 23:45:45.152351 +02:00] INFO [src\core_main.rs:162] main start args:["rustdesk:///"]
[2023-12-14 23:45:45.154136 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk.toml
[2023-12-14 23:45:45.154374 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk2.toml
[2023-12-14 23:45:45.277925 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk_local.toml
[2023-12-14 23:45:45.288304 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk_default.toml
[2023-12-14 23:45:47.202140 +02:00] INFO [libs\scrap\src\common\hwcodec.rs:382] Check hwcodec config, exit with: exit code: 0
[2023-12-14 23:45:59.344117 +02:00] ERROR [src\ui_interface.rs:1027] ipc connection closed: reset by the peer
[2023-12-14 23:46:03.851749 +02:00] ERROR [src\ui_interface.rs:1027] ipc connection closed: reset by the peer
[2023-12-14 23:46:20.665139 +02:00] INFO [src\platform\windows.rs:2151] Uninstalling service...
[2023-12-14 23:46:21.964457 +02:00] ERROR [src\ui_interface.rs:1027] ipc connection closed: reset by the peer
[2023-12-14 23:46:22.136564 +02:00] DEBUG [src\platform\windows.rs:2262] Spawn new window

---#--- RustDesk_rCURRENT.log
[2023-12-14 23:46:24.254232 +02:00] INFO [src\core_main.rs:162] main start args:["rustdesk:///"]
[2023-12-14 23:46:24.255756 +02:00] INFO [src\server.rs:476] server not started (will try to start): The system cannot find the file specified. (os error 2)
[2023-12-14 23:46:24.257894 +02:00] INFO [src\ipc.rs:278] Started ipc server at path: \\.\pipe\RustDesk\query
[2023-12-14 23:46:24.267242 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk2.toml
[2023-12-14 23:46:24.268365 +02:00] INFO [src\lan.rs:30] lan discovery listener started
[2023-12-14 23:46:24.384270 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk_local.toml
[2023-12-14 23:46:24.388047 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk.toml
[2023-12-14 23:46:24.393843 +02:00] DEBUG [libs\hbb_common\src\config.rs:491] Configuration path: C:\Users\theuser\AppData\Roaming\RustDesk\config\RustDesk_default.toml
```

Even if the error is not the same, the connection is recreated on mouse click/keyboard key, again and again.

**Comment by deep-soft** [comment]
Tested last 1.2.4 nightly on another Windows 10 22H2, clean install...
Same problems... connection closing every second.
Tried to increase timeout on (change timeout from 1s to 60s in processing loop), now connection is closed on mouse move/mouse click/keyboard press.
If I try to move some window, after reconnection the window is moved but again connection is closed, over and over.
:(
Reverted to https://github.com/rustdesk/rustdesk/commit/3f3bca753cc4a32706d052ca67d6fe531acf4989

**Reply by 21pages** [reply]
> Tried to increase timeout

Where is the timeout code, is it this?https://github.com/rustdesk/rustdesk/blob/853d655a92a456530eb60696e464999a73b00e17/src/server/connection.rs#L425

**Reply by deep-soft** [reply]
Yes, that one.

**Reply by 21pages** [reply]
Can you help with debugging? It seems that some bugs were introduced after that commit. You can print logs or commect out some code at the timeout or click event handling on the controlled end to see where the issue is originating from. Give this a try first and see if we can find any clues. 
https://github.com/21pages/rustdesk/releases/download/nightly/rustdesk-1.2.4-x86_64-install-disconnect.exe 
https://github.com/21pages/rustdesk/commit/167ed4d304d57511c738be2765b1b81a3e4c144f
`C:\Windows\ServiceProfiles\LocalService\AppData\Roaming\RustDesk\log\server\RustDesk_rCURRENT.log`

**Reply by deep-soft** [reply]
Forgot about this:
C:\Windows\ServiceProfiles\LocalService\AppData\Roaming\RustDesk\log\server\RustDesk_rCURRENT.log

Unfortunately I switch back to version from 11.19.
I'll try to build another version with changes you made (https://github.com/21pages/rustdesk/commit/167ed4d304d57511c738be2765b1b81a3e4c144f)
I can't run provided .exe, I must sign .exe with our own private key.

**Comment by lugt** [comment]
Thanks for the attention given, and to your hard work in making rustdesk more reliable than ever. 👍

**Comment by deep-soft** [comment]
> Can you help with debugging? It seems that some bugs were introduced after that commit. You can print logs or commect out some code at the timeout or click event handling on the controlled end to see where the issue is originating from. Give this a try first and see if we can find any clues. https://github.com/21pages/rustdesk/releases/download/nightly/rustdesk-1.2.4-x86_64-install-disconnect.exe [21pages@167ed4d](https://github.com/21pages/rustdesk/commit/167ed4d304d57511c738be2765b1b81a3e4c144f) `C:\Windows\ServiceProfiles\LocalService\AppData\Roaming\RustDesk\log\server\RustDesk_rCURRENT.log`

@21pages 
I made the changes, as here:
(https://github.com/21pages/rustdesk/commit/167ed4d304d57511c738be2765b1b81a3e4c144f)
I made new branch on my fork and tried to use you're commit on that branch, but didn't succeed.
I do not know how to add other upstream repository, as 21pages/rustdesk, same as rustdesk/rustdesk.
But I edited the two files and added the code from you're commit.
Here are the logs after the 1) installation, 2) start service and 3) first connection, for both user and LocalService.
[rustdesk-debug-logs-20231216-1931.zip](https://github.com/rustdesk/rustdesk/files/13693519/rustdesk-debug-logs-20231216-1931.zip)
I hope these will be helpful to you.

What I see after first connection:
- after mouse click/keyboard press the connection is closed (dropped?) and immediately recreated;
- the timeout (60 instead of 1) is not important in this case;
- after more than 120 seconds, with no mouse click or keyboard press, the connection is not closed;
- mouse move doesn't matter, connection is not closed;

**Reply by 21pages** [reply]
Thanks for your test. The log shows that the controlled side received mouse events continuously, but there is no log of disconnection. 
The disconnection occurs only when the controlled terminal is installed (while portable version is ok)? If both installed and portable failed, could yout provide the log of the controling side?

**Reply by deep-soft** [reply]
For portable version I need to run rustdesk.exe from `C:\Users\theuser\AppData\Local\rustdesk` ?

**Reply by 21pages** [reply]
Uninstall old one, click  rustdesk-1.2.4-x86_64.exe and run without install, and connect to it from another pc.

**Reply by deep-soft** [reply]
rustdesk.exe run from C:\Users\theuser\AppData\Local\rustdesk or from portable (rustdesk-1.2.4-x86_64.exe)

version 2023-12-16-dbg
service started, connection created to rustdesk network, but on first connection from remote the server is closed and connection from remote fails

version 2023-11-19
service started, connection NOT created to rustdesk network, can't connect from remote

In both version logs I see:
INFO [src\server.rs:476] server not started (will try to start): The system cannot find the file specified. (os error 2)
[log-04-portable.zip](https://github.com/rustdesk/rustdesk/files/13694620/log-04-portable.zip)

**Reply by 21pages** [reply]
So protable version also have this problem on your pc, `server not started (will try to start)` is normal. Could you upload the log of the control side? You can press server Enter on the log file before click mouse , and only upload the incremental log after the mosue click.

**Reply by deep-soft** [reply]
remote, portable, version 2023-11-19
 - used only for connection to remote
 - not installed
 - ip: xxx.yyy.zzz.xxx (in previous logs)
[log-05-portable-remote.zip](https://github.com/rustdesk/rustdesk/files/13694659/log-05-portable-remote.zip)

**Reply by 21pages** [reply]
You can press serveral Enters on the log file before click mouse , and only show the incremental log after the mouse click.

**Reply by deep-soft** [reply]
version 2023-12-16-dbg
- started as portable
- connection ok to rustdesk network
- waiting for connection

remote user
- connection to server
![image](https://github.com/rustdesk/rustdesk/assets/10506197/7151afdc-e090-4eb6-8ca3-9620a6ba409f)
- server is closed (does not apear in process list) on first attempt from remote user
- on remote user
![image](https://github.com/rustdesk/rustdesk/assets/10506197/1e7cdb49-dd29-42df-8935-4e6daa9267a1)

**Reply by 21pages** [reply]
> server is closed (does not apear in process list) on first attempt from remote user

What does this mean? When connected, the controlled portable version crashed?

**Reply by deep-soft** [reply]
yes

**Reply by deep-soft** [reply]
version 2023-12-16-dbg
I think controlled side crash on every connection, but if is installed the service restarts and receives from remote user clicks or keystrokes, apply them and crash again, over and over.
If portable, controlled side crash, and remote user can't connect because server is not active.

Strange!
version 2023-11-19
- on previous run the portable version does not accept connections, server was not started
- now, after I run the 2023-12-16-dbg, portable 2023-11-19 works! server is started and connection to rustdesk network is OK !

**Reply by deep-soft** [reply]
I think something is happening because of new Privacy Mode or/and IDD driver

**Reply by 21pages** [reply]
Could you confirm something for the controlled side crash:
`When`: mouse click or timer timeout
`The controlled side taskManager`: If controlled side installed, one rustdesk process will repeatedly crash and restart.
`The control side`: If controlled side installed, what connection error dialog does it show and then reconnect

**Comment by deep-soft** [comment]
`When:` mouse click or keyboard press, not on timer timeout
`The controlled side taskManager:` If controlled side installed, one rustdesk process will repeatedly crash and restart.
`The control side:` If controlled side installed

connection to server

server is closed (does not apear in process list) on first attempt from remote user
on remote user
can't capture the screen, but the message is 
Connection error
Reset by peer
Reconnect (1s) OK

**Reply by rustdesk** [reply]
Is this your fork or our nightly build?

**Reply by deep-soft** [reply]
fork from nightly build
- branch with changes from @21pages (debug messages)
- compilation on GH CI with custom server address and ports
- workflow based on flutter-build.yml (windows portion)

**Comment by fufesou** [comment]
@deep-soft 

I can hardly reproduce the crash.

Can you build the debug exe on your PC?
If you can please try the master commit?

1. Build the flutter version.
2. Copy the `target\debug\librustdesk.pdb` to `flutter\build\windows\runner\Debug`.
3. Optional. Replace the files in `flutter\build\windows\runner\Debug` with the files in `C:\Program Files\RustDesk` if you have installed RustDesk. Then restart the service.
3. Run the exe, reproduce the crash.
4. Package the rustdesk.dmp in `%LocalAppData%\CrashDumps` and all files in `flutter\build\windows\runner\Debug`.
5. Upload to somewhere.

I'll find the crash stack.

You can also skip step 1 and 2, and try my prebuilt package [Debug.zip](https://github.com/fufesou/rustdesk/releases/download/tmp/Debug.zip) in https://github.com/fufesou/rustdesk/releases/tag/tmp .

If you use my prebuilt package. You can only upload the rustdesk.dmp in `%LocalAppData%\CrashDumps`.

**Reply by deep-soft** [reply]
Thanks @fufesou for you're interest.
I'll try to use you're recommendations.

> Can you build the debug exe on your PC?

I can't successfully build on my home PC the flutter version, only the sciter one.
And also not on that PC, mentioned above, it's my work PC.

> If you can please try the master commit?

OK, i'll try to build the debug version on github CI.
On that work PC I can't run the nightly build or the prebuilt package above.

**Reply by rustdesk** [reply]
Please reproduce it with out your code change.

**Reply by fufesou** [reply]
> On that work PC I can't run the nightly build or the prebuilt package above.

Why? What error does it report? Or can you check the Windows event viewer?

![image](https://github.com/rustdesk/rustdesk/assets/13586388/b04544d2-7c11-4962-b741-465329097994)


> OK, i'll try to build the debug version on github CI.

github CI will build a release version, which does not contain debug symbols.
Make sure you have do some changes to build the debug version.

**Reply by deep-soft** [reply]
Based on build.py I made modifications to build Debug version.

**Reply by deep-soft** [reply]
Mistery solved !
At least my case.
From september, this year, I added to the workflow, 
{ target: x86_64-pc-windows-msvc, os: windows-2022 }
after existing
{ target: x86_64-pc-windows-msvc, os: windows-2019 }

For every built version I used the one compiled with "windows-2022" and until 19 november that was OK.
After that day, from 21 november, all versions compiled with "windows-2022" crashed on Windows 10 above 21H2.

Today, after many tries to build debug version with GH CI, I tried the release version compiled on 17 dec, on "windows-2019" and the crashes does NOT appeared !

Summary
After 2023.11.19, the versions built on Windows runner "windows-2022" crashed on Win10 above 21H2.

For this reason the nightly build from rustdesk/rustdesk (compiled on "windows-2019") does not crash and @21pages, @fufesou and @rustdesk can't reproduce the error.

Thank you for trying to help me to solve that, and sorry for the time you spent on this.

**Reply by rustdesk** [reply]
@deep-soft stop doing such thing next time (abuse our resource on your build), otherwise, I will block you.

**Reply by deep-soft** [reply]
Noted and understood.

**Comment by lugt** [comment]
My guess is that there is a problem in the installation process of Service Mode, apart from the one pointed by deep-soft. maybe someone else could post an issue later on.

**Comment by ethanriverpage** [comment]
Hey, having this same issue here. On my work laptop, so I assume it has something to do with GPO's defining security policies. @lugt , is this machine hooked up to a AD/AAD instance? Work/school laptop, etc.

**Comment by lugt** [comment]
This is my personal laptop, no AD that I am aware of, though I had a Office 365 account logged in(belonging to workplace), should that be of concern?

**Comment by scruel** [comment]
@rustdesk Same issue here, rustdesk just won't be useable because of this... I have to reboot the whole system every time for using it, and don't know what caused this.
Confirmed RpcSs service started, tried to run rustdesk with admin privileges, close the windows firewall, disable defender and related protections, nothing helped...

**Reply by scruel** [reply]
@rustdesk Tried the latest version and the issue is now gone in this version.

**Reply by rustdesk** [reply]
Thank you so much. Let me know if you have the issue again.

**Comment by thib3113** [comment]
Same issue here . I got a problem about keys on the server (bad docker volume) . When I debug my key problem, I reboot rustdesk app ... and it crash loop with the error `Failed to start ipc server at path \\.\pipe\RustDesk\query` . ( v1.2.4 ) . 

Reusing the correct keys on the server resolve the problem on my side

**Reply by rustdesk** [reply]
Can you still reproduce this issue constantly? Even with latest nightly https://github.com/rustdesk/rustdesk/releases/tag/nightly, I upgraded tokio (which fixed an ipc issue) yesterday, could you try?

**Reply by thib3113** [reply]
Ok, so I've done more tests with 1.2.4 version (not nightly) . 

First, moving key from server (so, the keys will be regenerated) . 

starting/killing the app => seems to doesn't reproduce

checking the rustdesk app, it say "ready", but can't connect because keys missmatch ( normal ) . 

I checked the windows services, and I found rustdesk service is off (strange that the app doesn't tell to start it ?) . 
So, I try to start it manually ... and it crash loop, and starting the rustdesk app will close it some seconds after . 
`C:\Windows\ServiceProfiles\LocalService\AppData\Roaming\RustDesk\log\server`
![image](https://github.com/rustdesk/rustdesk/assets/5489218/aad1c145-6f18-43e8-aa83-681b94eff8da)

---

So, I uninstall rustdesk, and try to start the install, but crash each times ... checking the task manager, it seems that a process of rustdesk is running

I kill the process, and restart the exe . It asks UAC ( not on the previous test ), and don't crash

---

So, installing the nightly : 

- the app say it's ready (but can't connect because keys missmatch ( normal )) => no crash
- try to reboot the app => no crash
- sometimes ask to start the service => service seems started, and doesn't crash loop
![image](https://github.com/rustdesk/rustdesk/assets/5489218/6d7e58b9-8295-40ae-ac57-6db8a47f88f1)


It seems fixed on nightly, I will update if I found errors later

--
( I restored the keys, and no errors )

**Reply by rustdesk** [reply]
Thank you so much.

**Comment by pfranzke** [comment]
I got a similar issue. I can't run the .exe as a user, but as an admin. The user has privileges to run all .exe programs. When I try to start Rustdesk, the loading symbol comes for a second and then nothing happens. I don't see any running process in Task Manager. I want to share Rustdesk as a portable app for my colleagues.

**Reply by rustdesk** [reply]
Your case is different from this thread. and obviously because of no IPC creation priviledge for normal user. Please show 
`C:\Users\<normal user>\AppData\Roaming\RustDesk\Logs\RustDesk_rCURRENT.log`

**Reply by pfranzke** [reply]
Unfortunately, there is no log from the newest version I tried to run. Previously, I used version 1.1.9, which stopped connecting to the other client's remote session. That's why I updated to the newest version, but this one won't even start and isn't writing any logs either. I have attached the old logs from version 1.1.9.[
[rustdesk-1.1.9_r2024-05-06_13-07-57.log](https://github.com/rustdesk/rustdesk/files/15306973/rustdesk-1.1.9_r2024-05-06_13-07-57.log)
[rustdesk-1.1.9_rCURRENT.log](https://github.com/rustdesk/rustdesk/files/15306974/rustdesk-1.1.9_rCURRENT.log)
[rustdesk-1.1.9_rCURRENT.log](https://github.com/rustdesk/rustdesk/files/15306976/rustdesk-1.1.9_rCURRENT.log)
](url)

**Reply by pfranzke** [reply]
Do you got any clue what the problem could be and what I should try?

**Reply by thib3113** [reply]
@pfranzke are you sure there is no rustdesk processes running ? ( check task manager ) ( I got this case )

**Reply by pfranzke** [reply]
@thib3113  Sadly not, I tried installing Rustdesk via "--silent-install" switch and this time the normal user is able to start Rustdesk, but I would like to  run it as portable version.

**Reply by rustdesk** [reply]
Could you try out if you can create named pipe with non-admin user as below?

https://serverfault.com/questions/1076043/how-to-add-permissions-to-windows-named-pipes-using-powershell

And I see this,  https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea#:~:text=Windows%C2%A010%2C%20version%201709%3A%C2%A0%C2%A0Pipes%20are%20only%20supported%20within%20an%20app%2Dcontainer%3B%20ie%2C%20from%20one%20UWP%20process%20to%20another%20UWP%20process%20that%27s%20part%20of%20the%20same%20app.%20Also%2C%20named%20pipes%20must%20use%20the%20syntax%20%5C%5C.%5Cpipe%5CLOCAL%5C%20for%20the%20pipe%20name. Not sure if related.

**Reply by pfranzke** [reply]
Sry, I never worked with pipes before. I tried some simple pipe with help from ChatGPT and ran it as normal user via PS. Idk if this is what you wanted? 
This was the code:

NamedPipeServer:
$pipeName = "testpipe"
$pipe = new-object System.IO.Pipes.NamedPipeServerStream($pipeName)
$pipe.WaitForConnection()
$reader = new-object System.IO.StreamReader($pipe)
$input = $reader.ReadLine()
Write-Output "Received message: $input"
$reader.Close()
$pipe.Close()

NamedPipeClient:
$pipeName = "testpipe"
$pipe = new-object System.IO.Pipes.NamedPipeClientStream(".", $pipeName, [System.IO.Pipes.PipeDirection]::Out)
$pipe.Connect()
$writer = new-object System.IO.StreamWriter($pipe)
$writer.AutoFlush = $true
$writer.WriteLine("Hello from PowerShell client!")
$writer.Close()
$pipe.Close()

**Reply by rustdesk** [reply]
Does above work? The  `NamedPipeServer` has output `Received message: Hello from PowerShell client!`?

**Reply by pfranzke** [reply]
Yes, the Server command waited until I ran the Client command, showed the output message and closed the pipe afterwards.

**Reply by rustdesk** [reply]
Thank you very much, our member @21pages will give you some test for finding out why our pipe creation failed with non-admin user.

**Reply by 21pages** [reply]
@pfranzke Could you run this program in terminal? https://github.com/21pages/test/releases/download/test/rustdesk_ipc_test.exe
This is the normal ouput.
```
[2024-05-16T09:47:45Z INFO  rustdesk] ipc start
[2024-05-16T09:47:45Z INFO  librustdesk::ipc] path: \\.\pipe\RustDesk\query_ipc_test
[2024-05-16T09:47:45Z INFO  librustdesk::ipc] before endpoint.incoming
[2024-05-16T09:47:45Z INFO  librustdesk::ipc] Started ipc_ipc_test server at path: \\.\pipe\RustDesk\query_ipc_test
[2024-05-16T09:47:46Z INFO  rustdesk] ipc connect success
[2024-05-16T09:47:46Z INFO  rustdesk] send success
```

**Reply by pfranzke** [reply]
I ran the .exe as local admin and normal user, but the ipc connect was not successfull.
![image](https://github.com/rustdesk/rustdesk/assets/169769151/a1f45b1c-e73a-4de9-84d7-55eef84c8f9d)

**Reply by pfranzke** [reply]
actually tried again and this time ran into error
![image](https://github.com/rustdesk/rustdesk/assets/169769151/4f156e41-a310-4c79-9e9f-898a1d7c75ee)

**Reply by 21pages** [reply]
@pfranzke   According to https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea#pipe_first_pipe_instance, FILE_FLAG_FIRST_PIPE_INSTANCE can return ERROR_ACCESS_DENIED (os error 5).
There is a possibility that the ipc is occupied and can't be start for the second time , could you verify that the second creation always fails?  https://github.com/21pages/test/releases/download/test/rustdesk_1st_arg_ipc_name.exe

This is normal output.
```
PS D:\rustdesk\rustdesk\rustdesk\target\debug> .\rustdesk_1st_arg_ipc_name.exe hello
[2024-05-16T11:34:58Z INFO  rustdesk] postfix: hello
[2024-05-16T11:34:58Z INFO  rustdesk] ipc start
[2024-05-16T11:34:58Z INFO  librustdesk::ipc] Started ipchello server at path: \\.\pipe\RustDesk\queryhello
[2024-05-16T11:34:59Z INFO  rustdesk] ipc connect success
[2024-05-16T11:34:59Z INFO  rustdesk] send success
PS D:\rustdesk\rustdesk\rustdesk\target\debug> .\rustdesk_1st_arg_ipc_name.exe hello
[2024-05-16T11:35:04Z INFO  rustdesk] postfix: hello
[2024-05-16T11:35:04Z INFO  rustdesk] ipc start
[2024-05-16T11:35:04Z INFO  librustdesk::ipc] Started ipchello server at path: \\.\pipe\RustDesk\queryhello
[2024-05-16T11:35:05Z INFO  rustdesk] ipc connect success
[2024-05-16T11:35:05Z INFO  rustdesk] send success
```

**Reply by pfranzke** [reply]
Yea the command worked.
![image](https://github.com/rustdesk/rustdesk/assets/169769151/6813a200-6f7c-4169-bf56-0e0b8d4faad0)

**Reply by 21pages** [reply]
@pfranzke  What if run it for the second time?

**Reply by pfranzke** [reply]
Working
![image](https://github.com/rustdesk/rustdesk/assets/169769151/1af1bda8-1502-4e49-81e5-be249d7d79ed)

**Reply by 21pages** [reply]
@pfranzke The second program exit gracely but the first is not, there is a possibility that an abnormal abort resulted in the pipe not being released. Could you run this program multiple times? https://github.com/21pages/test/releases/download/test/rustdesk_ipc_server_only_ctrl_c.exe
And could you run your powershell script multiple times? 
This is normal ouput.
```
PS D:\rustdesk\rustdesk\rustdesk\target\debug> .\rustdesk_ipc_server_only_ctrl_c.exe hello
[2024-05-17T02:08:08Z INFO  rustdesk] postfix: hello
[2024-05-17T02:08:08Z INFO  rustdesk] ipc start
[2024-05-17T02:08:08Z INFO  librustdesk::ipc] Started ipchello server at path: \\.\pipe\RustDesk\queryhello
PS D:\rustdesk\rustdesk\rustdesk\target\debug> .\rustdesk_ipc_server_only_ctrl_c.exe hello
[2024-05-17T02:08:12Z INFO  rustdesk] postfix: hello
[2024-05-17T02:08:12Z INFO  rustdesk] ipc start
[2024-05-17T02:08:12Z INFO  librustdesk::ipc] Started ipchello server at path: \\.\pipe\RustDesk\queryhello
```

**Reply by pfranzke** [reply]
My PS script is running without any problems. The new one you send me shows the normal output, but does not end the script. 
![image](https://github.com/rustdesk/rustdesk/assets/169769151/10b4ddb7-ebc0-4fd4-a3e3-b1be450d77d3)

**Reply by 21pages** [reply]
Yes, "does not end the script" is expected, because I want to reproduce the failed ipc like the first time.

**Reply by pfranzke** [reply]
I just recognized, when I am running the script already and try to run it in another cmd, the task will fail.
![image](https://github.com/rustdesk/rustdesk/assets/169769151/9bc25696-cbcd-435d-9032-97a22ee5dfe2)

**Reply by 21pages** [reply]
Yes, only one application is allowed to occupy the specified ipc.

**Reply by 21pages** [reply]
On your computer, is there any situation which can reproduce ipc failure with a high probability?

**Reply by pfranzke** [reply]
Not as far as I know.

**Comment by nate235** [comment]
Apologies for the "duplicate" I checked the issues section for this bug, not here...

**Reply by rustdesk** [reply]
No problem. Please try out https://github.com/rustdesk/rustdesk/discussions/6640#discussioncomment-9430285
It worked for @thib3113 and @scruel
